### PR TITLE
feat(benchmark): stats-loop tool-improvement issue dispatcher

### PR DESCRIPTION
**Pull request by jmoreira-valory**
## Summary

- Adds a **deterministic stats-loop** at the tail of the daily `benchmark-flywheel` workflow that reads per-platform rolling-window Brier + log-loss scores and opens a `tool-improvement`-labelled GitHub issue when a tool meets all gates on a Polymarket regression for two consecutive runs.
- Adds 13 unit tests covering each gate of the triage cascade.
- The opened issue is **data only** (regression headline, window definitions, machine-readable baseline stats, link to the run's `benchmark-data` artifact); no diagnosis or suggested fix. The investigating agent (`tool-improvement-agent` on the agent-skills monorepo VPS, dispatched via the `tool-improvement` label - see [agent-skills PR #146](https://github.com/valory-xyz/agent-skills/pull/146)) does the analysis.

## The cascade (§4 gates from the design)

All must pass for an issue to open:

| # | Gate | Rule | Why |
|---|---|---|---|
| 1 | Platform | Polymarket only | Omen Brier is confounded by `jury-resolve-market-v1` outcome judgement (R0014 §6, Grok-voter incident) |
| 2 | Regression size | Δ Brier > 0.015 | Matches the design's noise-vs-signal threshold |
| 3 | Sample floor | `valid_n ≥ 60` per window + soft per-day floor | Avoids triggering on holidays / post-cutover gaps |
| 4 | Sign agreement | sign(Δ Brier) == sign(Δ log_loss) | Two primary metrics moving the same way reduces FP rate |
| 5 | Reliability fork | `reliability < 0.80` → log warning, do NOT open issue | Reliability collapses are upstream (API outages, retired model slugs); LLM patching the tool source would paper over the symptom |
| 6 | Two-day confirmation | Today AND yesterday flagged the same tool | Cuts FP from ~15% to ~2% on the `n ∈ [60, 200]` band |

Cross-day state for gate 6 lives in `benchmark/results/stats_loop_state.json`; rides in the existing `benchmark-data` artifact across daily runs.

## What ships in this PR

```
benchmark/stats_loop/
├── __init__.py          - package marker + module docstring
├── triage.py            - gate cascade (pure data, no LLM, no network)
└── open_issue.py        - dispatcher entry point: gh issue create + state persist
benchmark/tests/test_stats_loop.py    - 13 tests
.github/workflows/benchmark_flywheel.yaml - new permissions, new triage step
```

Net diff: **+1159 LOC**, all new files except `benchmark_flywheel.yaml` (+35 LOC).

## Workflow changes

- New `permissions: { contents: read, issues: write }` block on the `benchmark` job. The default `github.token` only has `contents: read`; the new permission lets the triage step call `gh issue create`.
- New step `Tool-improvement triage (Polystrat)` at the tail of the `benchmark` job (after the per-platform Slack posts, before the daily-log archive upload). Defensive `hashFiles()` gate on both rolling-score files: if either is missing (because the scorer step failed), the triage step skips rather than opening a noisy half-data issue.
- `stats_loop_state.json` added to the `Upload benchmark data` step so the two-day confirmation state survives.

## Three new repo Variables (require manual config by the repo owner)

These trigger lint warnings in the IDE until they're declared at the repo-vars level — Pearl team / repo owner needs to set them in **Settings → Secrets and variables → Actions**:

| Variable | Default behavior | When to override |
|---|---|---|
| `ENABLE_TOOL_IMPROVEMENT` | enabled (any value other than `false` allows dispatch) | Set to `false` to silence dispatch while still computing + persisting state — recommended for the initial "observer mode" rollout (measure gate behavior for a week before the agent is live) |
| `TOOL_IMPROVEMENT_DRY_RUN` | not set | Set to `true` to log intended `gh issue create` calls without executing them — useful while watching the gate FP rate |
| `TOOL_IMPROVEMENT_GH_TOKEN` | `github.token` | Set to a dedicated bot PAT to author issues under that bot's identity rather than `github-actions[bot]` (optional) |

## Rollout recommendation

1. **Week 1 (observer mode):** set `ENABLE_TOOL_IMPROVEMENT=false` at the vars level. Triage runs every day, state file gets populated, but no issues open. Measure how often each gate would have fired, sanity-check the day-2 confirmation rate.
2. **Week 2 (dry-run):** set `ENABLE_TOOL_IMPROVEMENT=true`, `TOOL_IMPROVEMENT_DRY_RUN=true`. Issues would-be-opened are logged but not actually filed. Verify the body format renders correctly.
3. **Week 3 (live):** unset `TOOL_IMPROVEMENT_DRY_RUN`. Issues fire for real. The first few should be reviewed by a human before they're handed to the agent.

This phasing matches R0016 §7's P1 → P2 → P3 progression.

## Test plan

- [x] **`uv run pytest benchmark/tests/test_stats_loop.py`** — 13/13 pass locally.
- [ ] CI passes (the new tests, existing tests, lint pipeline).
- [ ] After merge + observer-mode rollout: spot-check that `stats_loop_state.json` populates with one entry per tool per day, and that two-day-confirmation logic fires correctly on a real regression.

## What's deliberately NOT in scope

- **The agent itself.** Lives in [agent-skills PR #146](https://github.com/valory-xyz/agent-skills/pull/146); separate review surface (Pearl team).
- **Installing the `valory-coding-agent` GitHub App on mech-predict.** Operational step; needs Pearl team coordination.
- **A new GitHub label.** `tool-improvement` is a soft label — GitHub auto-creates it on first use via `gh issue create --label tool-improvement`. Pearl team can optionally pre-create it with a color and description in Settings → Labels.
- **PR-CI's cached-replay re-bench** (the validation gate for the agent's draft PR). Already exists at `.github/workflows/benchmark_replay.yaml`; no changes needed.

## Test plan

- [x] 13/13 unit tests pass locally
- [ ] CI passes
- [ ] After merge: smoke-test by manually setting `TOOL_IMPROVEMENT_DRY_RUN=true` for one cycle and verifying log output

### CODE REVIEW

File: benchmark/stats_loop/triage.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+        yesterday = (
+            TriageOutcome(**yesterday_raw)
+            if isinstance(yesterday_raw, dict)
+            else None
+        )
+
+        # Gate 6 (confirmation): both today and yesterday must have flagged
+        # the same tool to open an issue. A first-day flag is silent (it
+        # just gets written to state for tomorrow to confirm).
+        if today.flagged and yesterday is not None and yesterday.flagged:
```
Comment (OjusWiZard) [inline benchmark/stats_loop/triage.py]: **[HIGH] Sustained regressions open a duplicate issue every day.**

`write_state` records `flagged=True` for an `open_issue` decision. On day N+1 the regression still flags today and yesterday's persisted state still says `flagged=True`, so this gate fires again → another `gh issue create`. There's no dedup against already-open `tool-improvement` issues and no "issue already opened" sentinel in state. A 5-day regression spams ~4 issues at the agent.

Fix: add an `issue_opened` flag to the state schema and suppress repeat `open_issue` (emit `silent` / `"already_open"`) until scores recover, **or** `gh issue list --label tool-improvement --search "<tool>"` before creating.

File: benchmark/stats_loop/triage.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+                    decision="reliability_collapse",
+                    today=collapse_outcome,
+                )
+            )
+            continue
+
+        today = _triage_one_tool(tool_name, cur_stats, prev_stats)
+        yesterday_raw = yesterday_by_tool.get(tool_name)
+        yesterday = (
+            TriageOutcome(**yesterday_raw)
```
Comment (OjusWiZard) [inline benchmark/stats_loop/triage.py]: **[HIGH] `TriageOutcome(**yesterday_raw)` crashes the whole triage loop on state-schema drift.**

`TriageOutcome` has 12 required fields, no defaults. The day a field is added/renamed, yesterday's persisted state (carried in the artifact) raises `TypeError`, `triage_tools` aborts, `write_state` never runs, and the **same crash recurs every subsequent run** until someone manually deletes the state file.

Fix: wrap in `try/except TypeError` → log and treat as `yesterday=None` (degrade to first-day confirmation). Optionally give optional fields `field(default=None)` for forward-compatible reads.

File: benchmark/stats_loop/open_issue.py [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+        "stats-loop summary: %d issue(s) opened, %d reliability collapse(s), "
+        "%d tool(s) silent",
+        n_opened,
+        n_reliability,
+        n_silent,
+    )
+
+    write_state(args.state, decisions, _now_iso())
+    log.info("State saved to %s", args.state)
+    return 0
```
Comment (OjusWiZard) [inline benchmark/stats_loop/open_issue.py]: **[HIGH] `gh issue create` failure is silently swallowed; `main()` always returns 0.**

A failed `gh` call (expired token, missing label, rate limit) is logged at ERROR but the step still exits green and the day's confirmed regression issue is lost with no CI signal.

Fix: track failures and `return 1 if n_failed else 0`. The workflow step has no `continue-on-error`, so a non-zero exit surfaces correctly — and state is already written before the return, so the next run can retry.

File: benchmark/stats_loop/open_issue.py [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+                repo=args.repo,
+                label=args.label,
+                title=title,
+                body=body,
+                dry_run=args.dry_run,
+            )
+            if rc == 0:
+                n_opened += 1
+        elif decision.decision == "reliability_collapse":
+            log.warning(
```
Comment (OjusWiZard) [inline benchmark/stats_loop/open_issue.py]: **[HIGH] `reliability_collapse` claims to "page a human" but only writes a log line.**

The docstrings (here and `triage.py`) say it pages an operator; the implementation is a single `log.warning`. If reliability collapses across tools (API outage / retired model slug), there is **zero human-visible alert** unless someone reads the Actions log — and zero issues open by design.

Fix: fire a real alert (reuse the Slack notify path with a `RELIABILITY_COLLAPSE` severity, or open a separate `reliability-collapse`-labelled issue), or correct the docstrings and add a runbook entry. Don't leave the doc/impl mismatch.

File: benchmark/tests/test_stats_loop.py [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+
+from __future__ import annotations
+
+import json
+from pathlib import Path
+from typing import Any, Dict
+
+import pytest
+
+from benchmark.stats_loop.triage import (
```
Comment (OjusWiZard) [inline benchmark/tests/test_stats_loop.py]: **[MEDIUM] `open_issue.py` has zero test coverage; several `triage.py` edge cases also untested.**

This suite imports only from `triage.py`. Untested in `open_issue.py`: `build_issue_body`, `build_issue_title` (parsed downstream by the agent — the format is a contract), `_window_iso_from_args` (W-2 disjointness/ordering), `_artifact_url`. Untested gate paths in `triage.py`: `reliability is None` (skips gate), `missing_brier` / `missing_log_loss`, the `60 ≤ valid_n < 105` sample-floor band, and `reliability == 0.80` boundary. Falls short of the repo's coverage bar.

File: benchmark/stats_loop/open_issue.py [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+    re-bench compares against. The keys MUST match what
+    ``benchmark/scorer.py`` writes into ``by_tool[<name>]``.
+    """
+    return json.dumps(stats, indent=2, sort_keys=True)
+
+
+def _artifact_url(run_id: str) -> str:
+    """Build the GitHub Actions UI URL for this run's artifacts page."""
+    return (
+        f"https://github.com/{DEFAULT_REPO}/actions/runs/{run_id}#artifacts"
```
Comment (OjusWiZard) [inline benchmark/stats_loop/open_issue.py]: **[MEDIUM] `_artifact_url` hardcodes `DEFAULT_REPO`, ignoring `--repo`.**

The issue is created on `args.repo` (`${{ github.repository }}`), but the artifact URL in the body always points to `valory-xyz/mech-predict`. Forks/staging produce issues with broken artifact links.

Fix: thread `args.repo` into `_artifact_url(run_id, repo)`.

File: benchmark/stats_loop/triage.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+    decision: str  # "open_issue" | "reliability_collapse" | "silent"
+    today: TriageOutcome
+    yesterday: Optional[TriageOutcome] = field(default=None)
+
+
+def _load_json(path: Path) -> Dict[str, Any]:
+    """Read a JSON file or return ``{}`` when the file is missing."""
+    if not path.exists():
+        return {}
+    return json.loads(path.read_text(encoding="utf-8"))
```
Comment (OjusWiZard) [inline benchmark/stats_loop/triage.py]: **[MEDIUM] Corrupt/partial JSON wedges the loop.**

`_load_json` only guards `not exists()`. A truncated state file (process killed mid-write) raises `JSONDecodeError`, which crashes the run; because state rides in the artifact, the corrupt write recurs every run.

Fix: catch `JSONDecodeError` here (log + treat as empty for the state file), and make `write_state` atomic (write `.tmp` then `os.replace`). See also the unguarded `json.loads` in `open_issue.py`.

**Review by OjusWiZard — COMMENTED** [review]
Multi-perspective review (code quality + test coverage + silent-failure hunt). Gate logic in `triage.py` is clean, well-documented, and well-tested; findings cluster in the dispatcher and operational robustness. Verified separately and **not** flagged: the scorer does emit the per-platform sibling files this code reads, and `stats_loop_state.json` round-trips correctly through the `benchmark-data` artifact.

**Headline:** 4 HIGH. The two I'd treat as merge-blockers are the duplicate-issue spam (H1) and swallowed `gh` failures (H3) — both directly undermine the feature. H2 and M3 are cheap insurance against the loop wedging itself. Inline comments below.

File: benchmark/stats_loop/open_issue.py [hunk 8]
```diff
@@ -0,0 +1,366 @@
...
+        prev_scores_path=args.prev_scores,
+        state_path=args.state,
+    )
+
+    now = datetime.now(timezone.utc)
+    window_iso = _window_iso_from_args(now)
+    artifact_url = _artifact_url(args.run_id) if args.run_id else "<no run id>"
+
+    polymarket_scores = (
+        json.loads(args.platform_scores.read_text(encoding="utf-8"))
```
Comment (OjusWiZard) [inline benchmark/stats_loop/open_issue.py]: **[MEDIUM] (same finding as the `_load_json` comment) Unguarded `json.loads` on the score files.**

A corrupt `scores_polymarket.json` / `scores.json` raises an unhandled `JSONDecodeError` and crashes `main()` with a raw traceback. Wrap both reads in `try/except json.JSONDecodeError` with explicit logging.

File: benchmark/stats_loop/triage.py [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+        return None
+    try:
+        return float(value)
+    except (TypeError, ValueError):
+        return None
+
+
+def _gate_sample_floor(stats: Dict[str, Any]) -> bool:
+    """Return True when the per-window sample floors are met."""
+    valid_n = stats.get("valid_n") or stats.get("n") or 0
```
Comment (OjusWiZard) [inline benchmark/stats_loop/triage.py]: **[MEDIUM] Sample-floor `valid_n or n` falls back to total `n` when `valid_n == 0`.**

`stats.get("valid_n") or stats.get("n") or 0` → a tool with 200 rows but 0 valid passes the floor on the total count, defeating the point of a *valid*-sample floor. Mostly masked today because the scorer sets `reliability = valid/total`, so `valid_n=0 ⇒ reliability 0 ⇒ reliability_collapse` short-circuits first — but it's a latent correctness smell (same pattern at line 180).

Fix: use `stats.get("valid_n", 0)` explicitly.

File: benchmark/stats_loop/triage.py [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+import json
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Any, Dict, List, Optional
+
+
+PLATFORM = "polymarket"
+BRIER_REGRESSION_THRESHOLD = 0.015
+N_WINDOW_FLOOR = 60
+N_DAY_FLOOR = 30
```
Comment (OjusWiZard) [inline benchmark/stats_loop/triage.py]: **[MEDIUM] `N_WINDOW_FLOOR = 60` is dead.**

Gate 2 in `_gate_sample_floor` (`valid_n / 7 < 15` ⇒ `valid_n < 105`) is always stricter than this 60-floor, so the constant never binds; the effective floor is 105. The test acknowledges this rather than fixing it. Set the constant to the real floor or drop the redundant check so the name and behaviour agree.

File: benchmark/stats_loop/open_issue.py [hunk 11]
```diff
@@ -0,0 +1,366 @@
...
+    """Compute the W-1 and W-2 ISO ranges based on the run wall-clock.
+
+    The scorer's per-window slicing uses calendar-day boundaries. This
+    approximation is good enough for issue-body context; the agent
+    re-derives the exact ranges from the daily log filenames when it
+    pulls the artifact.
+    """
+    fmt = "%Y-%m-%dT%H:%M:%SZ"
+    w1_end = now
+    w1_start = w1_end.replace() - _delta_days(days)
```
Comment (OjusWiZard) [inline benchmark/stats_loop/open_issue.py]: **[LOW] Dead `.replace()` no-op.**

`w1_end.replace()` with no args just copies the datetime. Use `w1_start = w1_end - timedelta(days=days)` and move `from datetime import timedelta` to the module-level imports (drop the deferred `_delta_days` helper).

File: benchmark/stats_loop/open_issue.py [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+    :return: Markdown issue body, ASCII-only.
+    """
+    today = decision.today
+    polymarket_block = _format_baseline_block(polymarket_stats)
+    combined_block = _format_baseline_block({
+        "tool": today.tool_name,
+        "stats": combined_stats,
+    })
+    return (
+        f"**Polystrat `{today.tool_name}`** Brier "
```
Comment (OjusWiZard) [inline benchmark/stats_loop/open_issue.py]: **[LOW] Implicit non-None contract.**

`{today.brier_cur:.4f}` / `brier_prev` / `delta_brier` are `Optional[float]`. The `open_issue` path guarantees they're set, but that's implicit — a future gate change would surface as a runtime `TypeError` (`None:.4f`) during a live run. Add an assert at the top of `build_issue_body` so it fails loudly.

File: benchmark/stats_loop/open_issue.py [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+            )
+            if rc == 0:
+                n_opened += 1
+        elif decision.decision == "reliability_collapse":
+            log.warning(
+                "RELIABILITY COLLAPSE on %s: reliability=%.3f (< %.2f). "
+                "Page an operator; tool-improvement issue NOT opened.",
+                decision.today.tool_name,
+                decision.today.reliability_cur or float("nan"),
+                0.80,
```
Comment (OjusWiZard) [inline benchmark/stats_loop/open_issue.py]: **[LOW] Magic `0.80` literal.** Use `triage.RELIABILITY_FLOOR` here instead of re-hardcoding the threshold.

File: benchmark/stats_loop/triage.py [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+tools have flagged a real regression worth opening an issue for.
+
+Gates (all must pass for a tool to be flagged):
+
+1. Platform: Polymarket only. Omen Brier is confounded by the on-chain
+   ``jury-resolve-market-v1`` mech tool's quality (see R0014 sec.6 and
+   the retired-Grok-voter incident); a tool-side fix cannot reliably
+   address an Omen regression, so the agent's scope excludes it.
+2. Regression size: ``Brier_cur - Brier_prev > 0.015``.
+3. Sample floor: ``n_window >= 60`` per window AND ``n_day >= 30``
```
Comment (OjusWiZard) [inline benchmark/stats_loop/triage.py]: **[LOW] Doc/impl mismatch.** This gate-3 docstring says `n_day >= 30`, but `_gate_sample_floor` softens to `N_DAY_FLOOR / 2` = 15/day. The inline comment there explains the softening; this top-level docstring should match.

File: benchmark/stats_loop/open_issue.py [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+
+    decisions = triage_tools(
+        cur_scores_path=args.cur_scores,
+        prev_scores_path=args.prev_scores,
+        state_path=args.state,
+    )
+
+    now = datetime.now(timezone.utc)
+    window_iso = _window_iso_from_args(now)
+    artifact_url = _artifact_url(args.run_id) if args.run_id else "<no run id>"
```
Comment (OjusWiZard) [inline benchmark/stats_loop/open_issue.py]: **[LOW] Silent run-id fallback.** When `GITHUB_RUN_ID` is unset, `"<no run id>"` is written into the issue body with no warning, and the downstream agent then mishandles it silently. Add a `log.warning` on this fallback (and document the env requirement for non-dry-run).

**Comment by jmoreira-valory** [comment]
Addressed all 15 findings in 3f6cd042. All 31 tests pass locally (`uv run pytest benchmark/tests/test_stats_loop.py -q`).

## HIGH (4)

| # | Fix |
|---|---|
| H1: duplicate-issue spam | Outcome now carries `issue_open` flag, persisted to state. Tomorrow's triage honors it and stays silent until gates 2-4 stop firing. New `TestDuplicateIssueSuppression` (3 tests) covers day-2-opens / day-3-suppresses / regression-resolves-clears. |
| H2: `TriageOutcome(**yesterday_raw)` crashes on schema drift | Replaced with `TriageOutcome.from_dict()`: unknown keys dropped, missing keys -> defaults, `TypeError/ValueError` caught. Worst case is "yesterday unknown" -> first-day-silent (self-heals next run). New `TestSchemaTolerantDeserialization` (4 tests). |
| H3: silent `gh issue create` failure | `open_github_issue` now returns `(rc, url)`. `main()` counts failures and exits non-zero so the workflow step turns red. State is still written before exit so tomorrow's confirmation gate isn't also wedged. Covered by `TestOpenIssueModule.test_open_github_issue_returns_failure_on_subprocess_error`. |
| H4: misleading docstring | Tightened both `triage.py` and `open_issue.py` docstrings to say what the code actually does (log WARNING; operator watching the workflow is the safety net). No functional change. |

## MEDIUM (5)

| # | Fix |
|---|---|
| M1: zero test coverage on `open_issue.py` | New `TestOpenIssueModule` with 9 tests covering `build_issue_title`, `build_issue_body` (including the L2 assert), `_artifact_url`, `_safe_load_json` (missing + corrupt), `_window_iso_from_args` (W-2 precedes W-1), and both `open_github_issue` paths (dry-run + subprocess failure). |
| M2: `_artifact_url` hardcoded `DEFAULT_REPO` | Now takes `repo` as a parameter; forks/staging get working artifact links. Covered by `test_artifact_url_uses_provided_repo`. |
| M3+M4: corrupt JSON wedges the loop | `_load_json` in `triage.py` and new `_safe_load_json` in `open_issue.py` both catch `(OSError, ValueError)`. New `TestCorruptStateFile` covers the state path; `TestOpenIssueModule` covers the score-file path. |
| M5: `valid_n or n` falls back to total n | New `_valid_n` helper reads strictly. A 200-row tool with 0 valid rows now correctly fails the sample floor. Covered by `TestValidNStrictness`. |
| M6: `N_WINDOW_FLOOR=60` dead | Constant dropped. Renamed to `N_DAY_FLOOR_PER_WINDOW_DAY=15` to make the per-day soft floor explicit. Module docstring + test docstrings updated to match. |

## LOW (5)

| # | Fix |
|---|---|
| L1: dead `.replace()` + deferred `_delta_days` | `_window_iso_from_args` rewritten; `timedelta` now a module-level import. Covered by `test_window_iso_w2_precedes_w1`. |
| L2: implicit non-None contract on `brier_cur/brier_prev/delta_brier` | Three asserts at the top of `build_issue_body`. Future gate change letting a None reach this path now fails loudly with a clear message instead of crashing mid-format. Covered by `test_build_issue_body_asserts_on_missing_brier`. |
| L3: magic 0.80 literal | Replaced with `triage.RELIABILITY_FLOOR`. |
| L4: doc/impl mismatch on `n_day >= 30` | Docstring corrected. |
| L5: silent `<no run id>` fallback | Now logs WARNING with the env-var name + a note that GitHub Actions provides it automatically. |

Test count: 13 -> 31. Ready for re-review.

**Comment by jmoreira-valory** [comment]
Closed in favour of #304
